java/maven_version.go

```diff
@@ -0,0 +1,301 @@
+		if len(a.List) == 0 {
+			return 0
+		}
+		for i := range a.List {
+			c := a.List[i].Compare(nil)
+			if c != 0 {
+				return c
+			}
+		}
+		return 0
+	case a.Kind() == listComponent && b.Kind() == listComponent:
+		for i := 0; i < len(a.List) || i < len(b.List); i++ {
+			var l, r *component
+			if i < len(a.List) {
+				l = &a.List[i]
+			}
+			if i < len(b.List) {
+				r = &b.List[i]
+			}
+			var res int
+			if l == nil {
+				if r != nil {
+					res = -1 * r.Compare(l)
+				}
+			} else {
+				res = l.Compare(r)
+			}
+			if res != 0 {
+				return res
+			}
+		}
+		return 0
+	case a.Kind() == listComponent && b.Kind() == intComponent:
+		return -1
+	case a.Kind() == listComponent && b.Kind() == stringComponent:
+		return 1
+	case a.Kind() == stringComponent && b == nil:
+		b = &component{Str: new(string)}
+		goto Again
+	case a.Kind() == stringComponent && b.Kind() == intComponent:
+		return -1
+	case a.Kind() == stringComponent && b.Kind() == listComponent:
+		return -1
+	default: // both strings
+		return strings.Compare(ordString(*a.Str), ordString(*b.Str))
+	}
+}
+
+func parseMavenVersion(s string) (*mavenVersion, error) {
+	v := &mavenVersion{
+		Orig: &s,
+	}
+	var b strings.Builder
+	l := &v.C.List
+	isDigit := false
+	pos := 0
+	for i, r := range s {
+		switch {
+		case r == '.':
+			if i == pos {
+				b.WriteByte('0')
+			}
+			if isDigit {
+				if err := appendInt(l, &b); err != nil {
+					return nil, err
+				}
+			} else {
+				appendString(l, &b)
+			}
+			pos = i + 1
+		case r == '-':
+			if i == pos {
+				b.WriteByte('0')
+			}
+			if isDigit {
+				if err := appendInt(l, &b); err != nil {
+					return nil, err
+				}
+			} else {
+				appendString(l, &b)
+			}
+			l = appendList(l)
+			pos = i + 1
+		case unicode.IsDigit(r):
+			if !isDigit && i > pos {
+				appendString(l, &b)
+				l = appendList(l)
+				pos = i
+			}
+			isDigit = true
+			b.WriteRune(r)
+		default:
+			if isDigit && i > pos {
+				if err := appendInt(l, &b); err != nil {
+					return nil, err
+				}
+				l = appendList(l)
+				pos = i
+			}
+			isDigit = false
+			b.WriteRune(r)
+		}
+	}
+	if isDigit {
+		if err := appendInt(l, &b); err != nil {
+			return nil, err
+		}
+	} else {
+		appendString(l, &b)
+	}
+	normalize(&v.C.List)
+	return v, nil
+}
+
+func appendInt(l *[]component, b *strings.Builder) error {
+	var v big.Int
+	if _, ok := v.SetString(b.String(), 10); !ok {
+		return fmt.Errorf("unable to parse number %q", b.String())
+	}
+	*l = append(*l, component{Int: &v})
+	b.Reset()
+	return nil
+}
+
+func appendString(l *[]component, b *strings.Builder) error {
+	s := b.String()
+	*l = append(*l, component{Str: &s})
+	b.Reset()
+	return nil
+}
+
+func appendList(l *[]component) *[]component {
+	ci := len(*l)
+	*l = append(*l, component{})
+	c := &(*l)[ci]
+	return &c.List
+}
+
+var zero = big.NewInt(0)
+
+func (c *component) isNull() bool {
+	return c == nil ||
+		(c.Int != nil && c.Int.Cmp(zero) == 0) ||
+		(c.Str != nil && *c.Str == "") ||
```

Comment:
I don't follow -- can you cook up an example version string?

Reply:
I believe this happens:

```
// Internally sets this.value to "" via this.value = ALIASES.getProperty(value, value);
// See https://github.com/apache/maven/blob/maven-3.9.x/maven-artifact/src/main/java/org/apache/maven/artifact/versioning/ComparableVersion.java#L346
Item release = new StringItem("release", false);
// I expect this to be "true" because it compares "" to "".
System.out.println(release.isNull());
```

Reply:
yeah, I just don't see how you'd get from a version string such that it matters. e.g. `1release` and `1` evaluate to being equal.